### PR TITLE
Add grafana_organization_user module

### PR DESCRIPTION
**Pull request by amenzhinsky**
##### SUMMARY

New module to manage user memberships and roles in Grafana organizations. 

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

 grafana_organization_user

### CODE REVIEW

**Comment by amenzhinsky** [comment]
Anyone?

**Comment by amenzhinsky** [comment]
@rrey @seuf could you review this PR?

File: tests/integration/targets/grafana_organization_user/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+---
+- name: Create test user
```
Comment (rrey) [inline tests/integration/targets/grafana_organization_user/tasks/main.yml:2]:
```suggestion
- name: Create test user (default org: 1)
```

File: tests/integration/targets/grafana_organization_user/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    url: "{{ grafana_url }}"
+    url_username: "{{ grafana_username }}"
+    url_password: "{{ grafana_password }}"
+    name: orgtest
+    email: orgtest@example.com
+    login: orgtest
+    password: userpassword
+    state: present
+
+- name: Remove user from organization
```
Comment (rrey) [inline tests/integration/targets/grafana_organization_user/tasks/main.yml:13]:
```suggestion
- name: Remove user from organization (default org: 1)
```

File: tests/integration/targets/grafana_organization_user/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    url_password: "{{ grafana_password }}"
+    login: orgtest
+    state: absent
+  register: result
+- assert:
+    that:
+      - "result.failed == false"
+      - "result.changed == true"
+
+- name: Check idempotency on user removal
```
Comment (rrey) [inline tests/integration/targets/grafana_organization_user/tasks/main.yml:26]:
```suggestion
- name: Check idempotency on user removal from org
```

File: tests/integration/targets/grafana_organization_user/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    url: "{{ grafana_url }}"
+    url_username: "{{ grafana_username }}"
+    url_password: "{{ grafana_password }}"
+    login: orgtest
+    state: absent
+  register: result
+- assert:
+    that:
+      - "result.failed == false"
+      - "result.changed == true"
```
Comment (rrey) [inline tests/integration/targets/grafana_organization_user/tasks/main.yml:24]: Can you also assert that the org and role returned are the expected one ? That will require that you add the user info in the module output.

File: plugins/modules/grafana_organization_user.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+DOCUMENTATION = '''
+---
+module: grafana_organization_user
+author:
+  - Aliaksandr Mianzhynski (@amenzhinsky)
+version_added: "1.5.0"
+short_description: Manage Grafana Organization Users.
+description:
+  - Add or remove users or change their roles in Grafana organizations through org API.
```
Comment (rrey) [inline plugins/modules/grafana_organization_user.py:30]: Can you specify if the user needs to exist before using this module ? and add a link to the grafana_user if they need to be created before using it ?

File: plugins/modules/grafana_organization_user.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+# Copyright: (c) 2021
+
+from __future__ import absolute_import, division, print_function
+
+DOCUMENTATION = '''
+---
+module: grafana_organization_user
+author:
+  - Aliaksandr Mianzhynski (@amenzhinsky)
+version_added: "1.5.0"
```
Comment (rrey) [inline plugins/modules/grafana_organization_user.py:27]: That would be 1.6.0

**Review by rrey — CHANGES_REQUESTED** [review]
Sorry for the long delay, but here is the review !

File: tests/integration/targets/grafana_organization_user/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    url_username: "{{ grafana_username }}"
+    url_password: "{{ grafana_password }}"
+    login: orgtest
+    role: editor
+    state: present
+  register: result
+- assert:
+    that:
+      - "result.failed == false"
+      - "result.changed == false"
```
Comment (rrey) [inline tests/integration/targets/grafana_organization_user/tasks/main.yml:79]: You don't have any test actually using the org_id input parameter. Could you add it in the coverage ?

**Comment by amenzhinsky** [comment]
@rrey all done, sorry for the delay.

File: plugins/modules/grafana_organization_user.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+# "avatarUrl": "/avatar/955c24be176f4feee34912a74de4ec01",
+# "email": "orgtest@example.com",
+# "lastSeenAt": "2012-07-29T13:42:00+04:00",
+# "lastSeenAtAge": "10 years",
+# "login": "orgtest",
+# "name": "orgtest",
+# "orgId": 1,
+# "role": "Editor",
+# "userId": 3
```
Comment (rrey-aviatrix) [inline plugins/modules/grafana_organization_user.py:92]: to remove

File: tests/integration/inventory [hunk 9]
```diff
@@ -0,0 +1,2 @@
+[testgroup]
```
Comment (rrey-aviatrix) [inline tests/integration/inventory:1]: this file shall be removed right ?

**Review by rrey-aviatrix — CHANGES_REQUESTED** [review]
@amenzhinsky you also commited pyc files that must be removed as well.

**Comment by amenzhinsky** [comment]
@rrey-aviatrix sorry for the delay, all done

**Comment by choooze** [comment]
Hello guys.

Any updates here? Looks like very useful module. For now there is no possibility to add user to any organization except of default one by ansible.

**Comment by amenzhinsky** [comment]
@rrey @seuf could you review this PR once again before it turns one year 😄

File: plugins/modules/grafana_organization_user.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    pass
+
+
+class GrafanaOrganizationUserInterface(object):
+
+    def __init__(self, module):
+        self._module = module
+        # {{{ Authentication header
+        self.headers = {"Content-Type": "application/json"}
+        self.headers["Authorization"] = basic_auth_header(module.params['url_username'], module.params['url_password'])
```
Comment (seuf) [inline plugins/modules/grafana_organization_user.py:149]: maybe you can add grafana_api_key usage (see https://github.com/ansible-collections/community.grafana/blob/main/plugins/modules/grafana_datasource.py#L673) for example.

Comment (rrey) [reply plugins/modules/grafana_organization_user.py:149]: @seuf the org api endpoint used do not support api token auth, see https://grafana.com/docs/grafana/latest/developers/http_api/org/#admin-organizations-api

It was the same for the users module

**Review by rrey — APPROVED** [review]

**Comment by rrey** [comment]
Thanks @amenzhinsky for the feature and test coverage. Sorry for the long delay for reviews.

**Comment by rrey** [comment]
@all-contributore add @amenzhinsky for test and code

**Comment by rrey** [comment]
@all-contributors add @amenzhinsky for code and test

**Pull request merged**